### ISSUE TITLE
ReferenceError: document is not defined when doing server rendering

### ISSUE THREAD

**Issue by kompot** [Yomguithereal/baobab-react#54]
Hi, I got the following when doing server render

```
/Users/kompot/projects/bm/front/node_modules/react/lib/getActiveElement.js:23
    return document.body;
           ^
ReferenceError: document is not defined
    at getActiveElement (/Users/kompot/projects/bm/front/node_modules/react/lib/getActiveElement.js:23:12)
    at ReactReconcileTransaction.ReactInputSelection.getSelectionInformation (/Users/kompot/projects/bm/front/node_modules/react/lib/ReactInputSelection.js:40:23)
    at ReactReconcileTransaction.Mixin.initializeAll (/Users/kompot/projects/bm/front/node_modules/react/lib/Transaction.js:168:30)
    at ReactReconcileTransaction.Mixin.perform (/Users/kompot/projects/bm/front/node_modules/react/lib/Transaction.js:133:12)
    at ReactUpdatesFlushTransaction.Mixin.perform (/Users/kompot/projects/bm/front/node_modules/react/lib/Transaction.js:134:20)
    at ReactUpdatesFlushTransaction.assign.perform (/Users/kompot/projects/bm/front/node_modules/react/lib/ReactUpdates.js:95:38)
    at Object.flushBatchedUpdates (/Users/kompot/projects/bm/front/node_modules/react/lib/ReactUpdates.js:175:19)
    at Object.wrapper [as flushBatchedUpdates] (/Users/kompot/projects/bm/front/node_modules/react/lib/ReactPerf.js:70:21)
    at ReactDefaultBatchingStrategyTransaction.Mixin.closeAll (/Users/kompot/projects/bm/front/node_modules/react/lib/Transaction.js:207:25)
    at ReactDefaultBatchingStrategyTransaction.Mixin.perform (/Users/kompot/projects/bm/front/node_modules/react/lib/Transaction.js:148:16)
    at Object.ReactDefaultBatchingStrategy.batchedUpdates (/Users/kompot/projects/bm/front/node_modules/react/lib/ReactDefaultBatchingStrategy.js:66:19)
    at Object.enqueueUpdate (/Users/kompot/projects/bm/front/node_modules/react/lib/ReactUpdates.js:215:22)
    at enqueueUpdate (/Users/kompot/projects/bm/front/node_modules/react/lib/ReactUpdateQueue.js:30:18)
    at Object.ReactUpdateQueue.enqueueSetState (/Users/kompot/projects/bm/front/node_modules/react/lib/ReactUpdateQueue.js:208:5)
    at ComposedComponent.ReactComponent.setState (/Users/kompot/projects/bm/front/node_modules/react/lib/ReactComponent.js:69:20)
    at ComposedComponent.<anonymous> (/Users/kompot/projects/bm/front/node_modules/baobab-react/dist-modules/higher-order.js:135:16)
```

`higher-order.js:135:16` is this line
https://github.com/Yomguithereal/baobab-react/blob/master/src/higher-order.js#L83

Found this (probably related) issue
https://github.com/facebook/react/issues/3620
in React repository.

Any thoughts how that could be fixed?

**Comment by Yomguithereal** [comment]
Hello @kompot. If I recall correctly, serverside rendering with `react` doesn't allow for component's lifecycle and state changes etc. This means you should provide a final tree to your app to render the desired state on the server before serving to the client. But I guess your use case is a bit more complicated. Are you performing data fecthing on the server the same way you do on the client?

For what it's worth, and I think this should be considered as a hack: [this](https://github.com/Yomguithereal/baobab-react/blob/master/test/endpoint.js#L26-L34) is how I use jsdom and make react think I can use the DOM when testing `baobab-react`.

Could you develop your use case a bit more here so we can advance on it together?

**Comment by nikhedonia** [comment]
I've encountered the same problem in my isomorphic app; my usecase is the following:

- Client has disabled js and client's application state is stored on the server.
- All possible state-transitions(actions) are encoded in urls.   

- When the User requests an url
- users app-state is selected from a cached baobab-tree.
- state mutations are derived from the url and performed 
- the react-router gets executed and the virtual dom(with baobab-context) created
- virtual dom gets stringified
- html-string is send to user
- on the second request, the described error shows up; i guess some cursor is still hooked to the virtual dom and forces the virtual dom to rerender

workaround was to create a temporary baobab:

```js
var session = new baobab( state.get(path) );
// do stuff
html = React.renderToString( <Root tree={session} />);

```

**Comment by kompot** [comment]
@Yomguithereal 
Here's the minimal app that has this issue
https://github.com/kompot/rjanko-forms/tree/217e16f748783b176626b83e5a69bfaafa0c8132/examples/users
you can checkout and try to run it.

Actual server rendering happens here
https://github.com/kompot/rjanko-forms/blob/217e16f748783b176626b83e5a69bfaafa0c8132/examples/users/src/renderApp.js#L24-L30
as you can see — if I modify json before creating Baobab tree then it (understandably) works ok,

Setting value on cursor after creating the tree — fails with 
```
 /Users/kompot/projects/rjanko-forms/examples/users/node_modules/react/lib/getActiveElement.js:23
     return document.body;
            ^
 ReferenceError: document is not defined
```
So nothing fancy there. Actually I could definitely make this work by preparing all the required JSON and initialize Baobab after that. But still would be great to understand.

Might this be somehow related to Baobab async updates?

**Comment by kompot** [comment]
In the mentioned React issue
https://github.com/facebook/react/issues/3620
it's stated that this should be fixed in React 0.14

Not closing this one, but watching closely ^)
